service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,209 @@
+import com.epam.esm.dao.creator.criteria.search.FullMatchSearchCertificateCriteria;
+import com.epam.esm.dao.creator.criteria.search.PartMatchSearchCertificateCriteria;
+import com.epam.esm.dao.creator.criteria.sort.FieldSortCertificateCriteria;
+import com.epam.esm.dto.GiftCertificate;
+import com.epam.esm.dto.Tag;
+import com.epam.esm.exception.InvalidFieldException;
+import com.epam.esm.exception.ResourceNotFoundException;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.TagService;
+import com.epam.esm.validator.TagValidator;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.collections4.SetUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.epam.esm.validator.GiftCertificateValidator.areGiftCertificateTagsValid;
+import static com.epam.esm.validator.GiftCertificateValidator.isDescriptionValid;
+import static com.epam.esm.validator.GiftCertificateValidator.isDurationValid;
+import static com.epam.esm.validator.GiftCertificateValidator.isGiftCertificateCreationFormValid;
+import static com.epam.esm.validator.GiftCertificateValidator.isNameValid;
+import static com.epam.esm.validator.GiftCertificateValidator.isPriceValid;
+
+/**
+ * The type Gift certificate service.
+ */
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService<GiftCertificate> {
+    private static final String ASC_SORT_ORDERING = "ASC";
+    private static final String DESC_SORT_ORDERING = "DESC";
+    private final GiftCertificateDao<GiftCertificate> dao;
+    private final TagService<Tag> tagService;
+
+    /**
+     * Instantiates a new Gift certificate service.
+     *
+     * @param dao        the dao
+     * @param tagService the tag service
+     */
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateDao<GiftCertificate> dao, TagService<Tag> tagService) {
+        this.dao = dao;
+        this.tagService = tagService;
+    }
+
+    @Override
+    public long insert(GiftCertificate giftCertificate) {
+        long id;
+        if (isGiftCertificateCreationFormValid(giftCertificate)) {
+            giftCertificate.setCreateDate(LocalDateTime.now());
+            if (!CollectionUtils.isEmpty(giftCertificate.getTags())) {
+                Set<Tag> existingTags = SetUtils.intersection(new HashSet<>(tagService.findAll(0, 0)),
+                        giftCertificate.getTags());
+                Set<Tag> newTags = new HashSet<>(CollectionUtils.removeAll(giftCertificate.getTags(), existingTags));
+                giftCertificate.setTags(newTags);
+                id = dao.insert(giftCertificate);
+                GiftCertificate certificateWithAllTags = dao.findById(id).get();
```

Comment:
what is the output of Objects.equals(certificateWithAllTags, giftCertificate)?

Reply:
False if the price is fractional otherwise true. I have an error in schema.sql (price decimal(10, 0) NOT NULL).